File: hk/hackathon/backend/app/engines/optimizer.py
"""Deterministic weighted-sum optimizer over 5 objectives (trimmed from the
7 described in the README's full spec: latency, cost, energy, capacity,
reliability). SLA acts as a hard epsilon-constraint, not just a scored
objective -- a breaching candidate can never outrank a satisfying one.
"""

WEIGHTS = {"latency": 0.30, "cost": 0.20, "energy": 0.15, "capacity": 0.20, "reliability": 0.15}


def normalize(values: dict, higher_is_better: bool) -> dict:
    lo, hi = min(values.values()), max(values.values())
    span = (hi - lo) or 1e-6
    out = {}
    for k, v in values.items():
        score = (v - lo) / span
        out[k] = score if higher_is_better else 1 - score
    return out


def rank_candidates(nodes: list[dict], predicted_latency: dict, sla_limit_ms: float) -> list[dict]:
    lat_norm = normalize({n["id"]: predicted_latency[n["id"]] for n in nodes}, higher_is_better=False)
    cost_norm = normalize({n["id"]: n["cost_per_hr"] for n in nodes}, higher_is_better=False)
    energy_norm = normalize({n["id"]: n["power_w"] for n in nodes}, higher_is_better=False)
    cap_norm = normalize({n["id"]: n["cpu_capacity"] - n["cpu_used"] for n in nodes}, higher_is_better=True)
    rel_norm = normalize({n["id"]: n["reliability"] for n in nodes}, higher_is_better=True)

    results = []
    for n in nodes:
        nid = n["id"]
        feasible = predicted_latency[nid] <= sla_limit_ms
        scores = {
            "latency": lat_norm[nid], "cost": cost_norm[nid], "energy": energy_norm[nid],
            "capacity": cap_norm[nid], "reliability": rel_norm[nid],
        }
        contributions = {k: round(scores[k] * WEIGHTS[k], 4) for k in scores}
        total = sum(contributions.values())
        results.append({
            "node_id": nid,
            "scores": contributions,
            "total": round(total, 4),
            "feasible": feasible,
            "reason_excluded": None if feasible else f"predicted {predicted_latency[nid]:.1f}ms exceeds SLA limit {sla_limit_ms:.0f}ms",
        })
    # feasible candidates always outrank infeasible ones, ties broken by total score
    results.sort(key=lambda r: (not r["feasible"], -r["total"]))
    return results

File: hk/hackathon/frontend-react/src/engines/optimizer.js
// Port of backend/app/engines/optimizer.py — same 5 objectives, same weights,
// same SLA epsilon-constraint rule. Not refactored, not reinvented.

export const WEIGHTS = { latency: 0.30, cost: 0.20, energy: 0.15, capacity: 0.20, reliability: 0.15 };

function normalize(valuesById, higherIsBetter) {
  const vals = Object.values(valuesById);
  const lo = Math.min(...vals), hi = Math.max(...vals);
  const span = (hi - lo) || 1e-6;
  const out = {};
  for (const [id, v] of Object.entries(valuesById)) {
    const score = (v - lo) / span;
    out[id] = higherIsBetter ? score : 1 - score;
  }
  return out;
}

/**
 * nodes: array of { id, cost_per_hr, power_w, cpu_capacity, cpu_used, reliability }
 * predictedLatency: { [nodeId]: ms }
 * slaLimitMs: number
 * Returns candidates ranked feasible-first, then by total score descending —
 * a candidate predicted to breach the SLA can never outrank one that satisfies it.
 */
export function rankCandidates(nodes, predictedLatency, slaLimitMs) {
  const latById = {}, costById = {}, energyById = {}, capById = {}, relById = {};
  for (const n of nodes) {
    latById[n.id] = predictedLatency[n.id];
    costById[n.id] = n.cost_per_hr;
    energyById[n.id] = n.power_w;
    capById[n.id] = n.cpu_capacity - n.cpu_used;
    relById[n.id] = n.reliability;
  }
  const latNorm = normalize(latById, false);
  const costNorm = normalize(costById, false);
  const energyNorm = normalize(energyById, false);
  const capNorm = normalize(capById, true);
  const relNorm = normalize(relById, true);

  const results = nodes.map((n) => {
    const id = n.id;
    const feasible = predictedLatency[id] <= slaLimitMs;
    const scores = {
      latency: latNorm[id], cost: costNorm[id], energy: energyNorm[id],
      capacity: capNorm[id], reliability: relNorm[id],
    };
    const contributions = {};
    for (const k of Object.keys(scores)) contributions[k] = Number((scores[k] * WEIGHTS[k]).toFixed(4));
    const total = Number(Object.values(contributions).reduce((a, b) => a + b, 0).toFixed(4));
    return {
      node_id: id,
      scores: contributions,
      total,
      feasible,
      reason_excluded: feasible ? null : `predicted ${predictedLatency[id].toFixed(1)}ms exceeds SLA limit ${slaLimitMs.toFixed(0)}ms`,
    };
  });

  results.sort((a, b) => {
    if (a.feasible !== b.feasible) return a.feasible ? -1 : 1;
    return b.total - a.total;
  });
  return results;
}
